components/brave_wallet/common/eth_abi_utils.h

```diff
@@ -81,6 +82,72 @@ class TupleEncoder {
   std::vector<Element> elements_;
 };
 
+enum class TypeKind {
+  // bytes<M> where 0 < M <= 32
+  kBytesM,
+
+  kAddress,
+
+  // uint<M> where 0 < M <= 256 and M % 8 == 0
+  // uint is an alias for uint256.
+  kUintM,
+
+  kBool,
+
+  kBytes,
+  kString,
+  kArray,
+  kTuple,
+};
+
+struct Type {
+  TypeKind kind;
+
+  // Indicates bit length for fixed-size types if applicable.
+  std::optional<size_t> m;
+
+  // Indicates the type of the array elements, if kind is kArray.
+  std::unique_ptr<Type> array_type;
+
+  // Indicates the types of the tuple elements, if kind is kTuple.
+  std::vector<Type> tuple_types;
+
+  explicit Type(TypeKind kind);
+  Type(TypeKind kind, size_t m);
+  ~Type();
+
+  // Delete copy constructor and copy assignment operator
+  Type(const Type&) = delete;
+  Type& operator=(const Type&) = delete;
+
+  // Define move constructor
+  Type(Type&& other) noexcept;
+};
```

Comment:
these kinds of methods typically come first

Reply:
53382d28e24530454970d1c2c9430dcc47c47f76